### PR TITLE
set a minimal C++ standard

### PR DESCRIPTION
**Pull request by nim65s**
Hi,

This PR defines a `CHECK_MINIMAL_CXX_STANDARD()` macro.

This allows to automatically set `CMAKE_CXX_STANDARD` if it is not already manually defined somewhere else (on an equal or greater version), and if we know that we require a standard version higher than the default one of the current compiler.

This can avoid to put `set(CMAKE_CXX_STANDARD 11)` in a `CMakeLists.txt` because C++11 features are required, even if more and more compilers will default to C++14 and more.

Also, when a package doesn't actually require any particular standard, but one of its dependency has been compiled with C++11, it's important to also compile in C++11. For this to work, we just need to call this macro from the generated *Config.cmake.

### CODE REVIEW

File: cxx-standard.cmake [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    # Check that the requested minimum is higher than the compiler default
+    # ref https://en.cppreference.com/w/cpp/preprocessor/replace#Predefined_macros for constants
+    if(_COMPILER_DEFAULT_CXX_STANDARD EQUAL 199711
+        OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201103 AND _MINIMAL_CXX_STANDARD GREATER 11)
+        OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201402 AND _MINIMAL_CXX_STANDARD GREATER 14)
+        OR (_COMPILER_DEFAULT_CXX_STANDARD EQUAL 201703 AND _MINIMAL_CXX_STANDARD GREATER 17))
+      # Check that the requested minimum is higher than any pre-existing CMAKE_CXX_STANDARD
+      if(NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD EQUAL 98 OR CMAKE_CXX_STANDARD LESS _MINIMAL_CXX_STANDARD)
+        set(CMAKE_CXX_STANDARD ${STANDARD})
+        message(STATUS "CMAKE_CXX_STANDARD upgraded to ${CMAKE_CXX_STANDARD}")
```
Comment (jmirabel) [inline cxx-standard.cmake:50]: Shouldn't it fail when CMAKE_CXX_STANDARD is defined and strictly less than _MINIMAL_CXX_STANDARD ?

Comment (nim65s) [reply cxx-standard.cmake:50]: This depend on what we want it to do. Maybe something more than a STATUS would be useful. 

But as it is now, if one call `SET_MINIMAL_CXX_STANDARD(11)` early in the `CMakeLists.txt, and then in a dependency Config.cmake file there is a `SET_MINIMAL_CXX_STANDARD(14)`, the `CMAKE_CXX_STANDARD`is upgraded twice. So it shouldn't fail in this case.

Comment (nim65s) [reply cxx-standard.cmake:50]: We could add a guard, like a `KEEP_CMAKE_CXX_STANDARD`, which would change lines 49 & 50 to a `FATAL_ERROR` message.

**Comment by nim65s** [comment]
We should totally do something like this.

I didn't though about this `if(COMMAND`,  so I was going to open another issue to discuss how we should do stuff like those. But this `if` makes my issue orthogonal to this feature, so let's do it :)

File: cxx-standard.cmake (hunk 1 above)
Comment (jmirabel) [reply cxx-standard.cmake:50]: > But as it is now, if one call `SET_MINIMAL_CXX_STANDARD(11)` early in the `CMakeLists.txt`, and then in a dependency Config.cmake file there is a `SET_MINIMAL_CXX_STANDARD(14)`, the `CMAKE_CXX_STANDARD`is upgraded twice. So it shouldn't fail in this case.

You are right.

**Comment by gergondet** [comment]
That is useful indeed, thanks @nim65s 

> We should totally do something like this.
> 
> I didn't think about this `if(COMMAND`, so I was going to open another issue to discuss how we should do stuff like those. But this `if` makes my issue orthogonal to this feature, so let's do it :)

You could also ship `cxx-standard.cmake` and `cxx-standard.cpp` along the `*Config.cmake` so that the command can become available if the project is consumed from a project that doesn't use jrl-cmakemodules

In general, I think the best approach to this problem is to use [`target_compile_features`](https://stackoverflow.com/questions/29687331/how-to-define-transitive-cxx-standard-c11-in-cmake)  but I found that not very reliable on older CMake versions unfortunately :(

**Comment by nim65s** [comment]
Shipping stuff like `cxx-standard.cmake` and `cxx-standard.cpp` is exactly what I would like to discuss in a separate issue :)

I agree that `target_compile_features` would be cleaner, but it generally requires more work, that I don't usually think it's worth it.

**Comment by jmirabel** [comment]
The only thing which annoys me with putting this in the dependency is that is one writes:
```cmake
set(CMAKE_CXX_STANDARD 11)
find_package(Foo)
```
and that `FooConfig.cmake` does `SET_MINIMUM_CXX_STANDARD(14)`, then, you won't have an error. There is still a message in the logs, but it isn't very clear. I think the dependencies should only check and never set `CMAKE_CXX_STANDARD`.

**Comment by jmirabel** [comment]
Maybe we could add an argument to `SET_MINIMUM_CXX_STANDARD` that enable to switch between those two behaviors:
1. update the minimum C++ standard and **set** CMAKE_CXX_STANDARD if necessary,
2. update the minimum C++ standard and **check** the CMAKE_CXX_STANDARD.

In the dependency, I think we only want 2 while in the package itself, we want 1.

**Comment by nim65s** [comment]
I understand your point, but one of my goal was to completely get rid of tuning robotpkg for C++ versions. For this, I would need that in dependencies, we have 1.

We could compromise by having 1 in the package and 2 in the dependencies, plus another flag that would enforce 2 to upgrade to 1, no matter where, like a manual `-DENFORCE_MINIMAL_CXX_STANDARD=ON`. I'll give that a try.

**Comment by nim65s** [comment]
Done. This is ready for reviews again :)

**Comment by traversaro** [comment]
Hi @nim65s , some questions as a "user" of packages that itself use `jrl-cmakemodules`: 
* Just to understand, what is the rationale of using this infrastructure instead of the usual `target_compile_features` and its meta-features `cxx_std_11`, `cxx_std_14`, .. ? 
* The use of `try_run` basically breaks any cross-compilation build. While as far as I understand this can be avoided by defining `_COMPILER_DEFAULT_CXX_STANDARD` , that is not documented and I guess is an internal detail that may change in the future. It would be nice if there was a documented way to disable the `try_run` call, to make sure that it can be set for cross compilation builds.

**Comment by nim65s** [comment]
@jmirabel : done, thanks !

@traversaro : Good questions, thank you :)

For the first one, `cxx_std_11` [is not available in CMake 3.1](https://cmake.org/cmake/help/v3.1/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html), and we still support 16.04 with its default 3.1 CMake. There also are other cases, when it's easier to define something at the project level, instead of the target level. But the possibilities added by this PR should not interfere in an incorrect way with a correct use of `cxx_std_11`.

For the second one, I have improved the documentation.

**Comment by traversaro** [comment]
Thank @nim65s !

**Comment by traversaro** [comment]
By the way, it is not a big problem but I am afraid the default C++ detection logic is not going on MSVC, see https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/ . TL;DR: In MSVC to get a meaningful value of `__cplusplus` you need to pass the `/Zc:__cplusplus` compile option.

**Comment by nim65s** [comment]
@traversaro : thanks for this remark. It looks like 85089a9 fixed that:
- before: [-- currently used  C++ standard: 199711](https://github.com/nim65s/example-adder/runs/944101231?check_suite_focus=true)
- after: [-- currently used  C++ standard: 201402](https://github.com/nim65s/example-adder/runs/944157918?check_suite_focus=true)

File: cxx-standard.cmake [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+macro(CHECK_MINIMAL_CXX_STANDARD STANDARD)
+  set(options ENFORCE)
+  set(oneValueArgs)
+  set(multiValueArgs)
+  cmake_parse_arguments(MINIMAL_CXX_STANDARD "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+  # Get compiler default cxx standard, by printing "__cplusplus" (only once)
+  if(NOT DEFINED _COMPILER_DEFAULT_CXX_STANDARD)
+    if(MSVC)
+      string(APPEND CMAKE_CXX_FLAGS " /Zc:__cplusplus")
```
Comment (jmirabel) [inline cxx-standard.cmake:43]:
```suggestion
      # See https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
      string(APPEND CMAKE_CXX_FLAGS " /Zc:__cplusplus")
```

**Comment by jmirabel** [comment]
From my side, this can be merged. Thank you @nim65s

**Pull request merged**